primitives/xcm/src/lib.rs

```diff
@@ -59,68 +59,88 @@ impl<
 	}
 }
 
+pub trait DropAssetsWeigher {
+	fn fungible() -> u64;
+	fn native() -> u64;
+	fn default() -> u64;
+}
+
 pub struct TrappistDropAssets<
 	AssetId,
 	AssetIdInfoGetter,
 	AssetsPallet,
 	BalancesPallet,
 	XcmPallet,
-	AccoundId,
->(PhantomData<(AssetId, AssetIdInfoGetter, AssetsPallet, BalancesPallet, XcmPallet, AccoundId)>);
-impl<AssetId, AssetIdInfoGetter, AssetsPallet, BalancesPallet, XcmPallet, AccountId> DropAssets
+	AccountId,
+	Weigher,
+>(
+	PhantomData<(
+		AssetId,
+		AssetIdInfoGetter,
+		AssetsPallet,
+		BalancesPallet,
+		XcmPallet,
+		AccountId,
+		Weigher,
+	)>,
+);
+
+impl<AssetId, AssetIdInfoGetter, AssetsPallet, BalancesPallet, XcmPallet, AccountId, Weigher>
+	DropAssets
 	for TrappistDropAssets<
 		AssetId,
 		AssetIdInfoGetter,
 		AssetsPallet,
 		BalancesPallet,
 		XcmPallet,
 		AccountId,
+		Weigher,
 	> where
-	AssetId: Clone,
 	AssetIdInfoGetter: AssetMultiLocationGetter<AssetId>,
 	AssetsPallet: Inspect<AccountId, AssetId = AssetId>,
 	BalancesPallet: Currency<AccountId>,
 	XcmPallet: DropAssets,
+	Weigher: DropAssetsWeigher,
 {
 	// assets are whatever the Holding Register had when XCVM halts
-	fn drop_assets(origin: &MultiLocation, assets: Assets) -> u64 {
-		let multi_assets: Vec<MultiAsset> = assets.into();
-		let mut trap: Vec<MultiAsset> = Vec::new();
+	fn drop_assets(origin: &MultiLocation, mut assets: Assets) -> u64 {
+		const NATIVE_LOCATION: MultiLocation = MultiLocation { parents: 0, interior: Here };
 
-		for asset in multi_assets {
-			if let MultiAsset { id: Concrete(location), fun: Fungible(amount) } = asset.clone() {
-				// is location a fungible on AssetRegistry?
-				if let Some(asset_id) = AssetIdInfoGetter::get_asset_id(location.clone()) {
-					let min_balance = AssetsPallet::minimum_balance(asset_id);
+		let mut weight = {
+			assets.non_fungible.clear();
+			Weigher::default()
+		};
 
-					// only trap if amount ≥ min_balance
-					// do nothing otherwise (asset is lost)
-					if min_balance <= amount.saturated_into::<AssetsPallet::Balance>() {
-						trap.push(asset);
-					}
+		assets.fungible.retain(|id, &mut amount| {
+			if let Concrete(location) = id {
+				match AssetIdInfoGetter::get_asset_id(location) {
+					Some(asset_id) => {
+						weight += Weigher::fungible();
```

Comment:
yep, I was unsure how it will work on weights v2. Weights v2 don't implement `Saturating` trait. I cheked now and it seems that the api remains the same, so I fixed it now.
thanks for pointing that out.